pkg/api/helpers/upload.go

```diff
@@ -86,68 +99,260 @@ func ClientUpload(ctx context.Context, client apigen.ClientWithResponsesInterfac
+		parts[i].URL = (*mpu.PresignedUrls)[i]
+		rem -= size
+		off += size
+	}
+
+	// upload parts in parallel, fill uploadParts array with results
+	uploadParts := make([]apigen.UploadPart, u.numParts)
+	g, grpCtx := errgroup.WithContext(context.Background())
+	g.SetLimit(u.uploader.Concurrency)
+
+	for i := 0; i < u.numParts; i++ {
+		i := i // pinning
+		g.Go(func() error {
+			etag, err := u.uploadPart(grpCtx, parts[i].Reader, parts[i].URL)
+			if err != nil {
+				return fmt.Errorf("part %d %w", i+1, err)
+			}
+			uploadParts[i] = apigen.UploadPart{
+				PartNumber: i + 1,
+				Etag:       etag,
+			}
+			return nil
+		})
+	}
+
+	// wait for all parts to be uploaded
+	if err := g.Wait(); err != nil {
+		// abort upload using a new context to avoid context cancellation
+		abortErr := u.abortMultipart(context.Background(), mpu)
+		if abortErr != nil {
+			err = errors.Join(err, abortErr)
 		}
-		return stats, nil
+		return nil, err
 	}
+
+	return u.completeMultipart(ctx, mpu, uploadParts)
 }
 
-func isAzureBlobURL(u string) bool {
-	parsedURL, err := url.Parse(u)
+func (u *presignUpload) completeMultipart(ctx context.Context, mpu *apigen.PresignMultipartUpload, uploadParts []apigen.UploadPart) (*apigen.ObjectStats, error) {
+	resp, err := u.uploader.Client.CompletePresignMultipartUploadWithResponse(ctx, u.repoID, u.branchID, mpu.UploadId,
+		&apigen.CompletePresignMultipartUploadParams{
+			Path: u.objectPath,
+		},
+		apigen.CompletePresignMultipartUploadJSONRequestBody{
+			Parts:           uploadParts,
+			PhysicalAddress: mpu.PhysicalAddress,
+			UserMetadata: &apigen.CompletePresignMultipartUpload_UserMetadata{
+				AdditionalProperties: u.metadata,
+			},
+			ContentType: apiutil.Ptr(u.contentType),
+		},
+	)
 	if err != nil {
-		return false
+		return nil, fmt.Errorf("complete presign multipart upload: %w", err)
 	}
-	return strings.HasSuffix(parsedURL.Host, ".blob.core.windows.net")
+	if resp.JSON409 != nil {
+		return nil, ErrConflict
+	}
+	if resp.JSON200 == nil {
+		return nil, fmt.Errorf("complete presign multipart upload: %w", ResponseAsError(resp))
+	}
+	return resp.JSON200, nil
+}
+
+func (u *presignUpload) abortMultipart(ctx context.Context, mpu *apigen.PresignMultipartUpload) error {
+	resp, err := u.uploader.Client.AbortPresignMultipartUploadWithResponse(ctx, u.repoID, u.branchID, mpu.UploadId,
+		&apigen.AbortPresignMultipartUploadParams{
+			Path: u.objectPath,
+		},
+		apigen.AbortPresignMultipartUploadJSONRequestBody{
+			PhysicalAddress: mpu.PhysicalAddress,
+		})
+	if err != nil {
+		return fmt.Errorf("abort presign multipart upload: %w", err)
+	}
+	if resp.StatusCode() != http.StatusNoContent {
+		return fmt.Errorf("abort presign multipart upload: %w", ResponseAsError(resp))
+	}
+	return nil
+}
+
+func (u *presignUpload) initMultipart(ctx context.Context) (*apigen.PresignMultipartUpload, error) {
+	// adjust part size
+	u.partSize = DefaultUploadPartSize
+	if u.size/u.partSize >= int64(MaxUploadParts) {
```

Comment:
This is kind of awkward. I think we should enforce that at least (MaxUploadParts + x) * u.partSize > u.size, where x is a constant size.
Otherwise, this can lead to bad performance

Reply:
I've used the same logic the s3 sdk uses to calculate the part size the client will use. It uses default part size and try to use it until the max number of parts allowed.
In the case the size can't fit the parts it will adjust the size.
The cli usually gives a way to configure the part size, but for now I kept it as the default which is the min size.